group/assignment5/UnknownAlphabet.java

```diff
@@ -83,10 +87,126 @@ public boolean isConsistentDictionary(List<String> dictionary) {
+    Graph.Result detectCycleResult = alphabetGraph.detectCycle();
+    while (detectCycleResult.hasCycle()) {
+      List<Vertex> cyclePair = detectCycleResult.getPair();
+      dictionary = removeInconsistency(cyclePair, dictionary, alphabetGraph);
+      detectCycleResult = alphabetGraph.detectCycle();
+    }
+
+    return dictionary;
+  }
+
+  /**
+   * Removes the minimum number of words from an inconsistent dictionary to make it consistent.
+   *
+   * Time complexity: O(n*m), where n is the number of words in the dictionary and m is the average length of a word.
+   *
+   * @param cyclePair the vertices that form the edge that needs to be removed
+   * @param dictionary the inconsistent dictionary of interest
+   * @param graph the graph created from the given dictionary
+   * @return a consistent dictionary that is a maximal subset of the inconsistent dictionary
+   */
+  private List<String> removeInconsistency(List<Vertex> cyclePair, List<String> dictionary, Graph graph) {
+    // the vertices that form the edge that creates the cycle
+    Vertex firstVertex = cyclePair.get(0);
+    Vertex secondVertex = cyclePair.get(1);
+    // dictionary without words imposing rule: firstVertex before secondVertex
+    List<String> firstNewDictionary = filterDictionaryWords(firstVertex, secondVertex, dictionary);
+    // dictionary without words imposing rule: secondVertex before firstVertex
+    List<String> secondNewDictionary = filterDictionaryWords(secondVertex, firstVertex, dictionary);
+
+    boolean moreWordsInFirstDictionary =  firstNewDictionary.size() > secondNewDictionary.size();
+    List<String> newDictionary = moreWordsInFirstDictionary? firstNewDictionary : secondNewDictionary;
+
+    if(moreWordsInFirstDictionary) {
+      // the constraint where the firstVertex is before secondVertex should be removed
+      graph.getVertex(firstVertex.getValue()).removeNeighbour(secondVertex);
+    } else {
+      // the constraint where the secondVertex is before firstVertex should be removed
+      graph.getVertex(secondVertex.getValue()).removeNeighbour(firstVertex);
+    }
+
+    return newDictionary;
+  }
+
+  /**
+   * Creates a new dictionary by removing the words that impose the lexicographic order of a given
+   * letter before another given letter
+   *
+   * Complexity: O(m*n^2), where n is the number of words in the dictionary, m is the average length of a word
+   * (n-1) * m + (n-2) * m + ... + 2 * m + m = m * ((n-1) + (n-2) + ... + 2 + 1) = m * (n^2)
+   *
+   * @param firstVertex vertex corresponding to first letter
+   * @param secondVertex vertex corresponding to second letter
+   * @param dictionary the dictionary of interest
+   * @return the given dictionary without the words that impose the rule where first letter is before
+   * second letter
+   */
+  private List<String> filterDictionaryWords(Vertex firstVertex, Vertex secondVertex, List<String> dictionary) {
+    Set<String> newDictionary = new LinkedHashSet<>();
```

Comment:
what is the difference btwn HashSet and LinkedHashSet? is there a reason for picking this one ?

Reply:
The LinkedHashSet maintains the insertion order of the elements. I used the LinkedHashSet to make sure that every word is added only once in the dictionary and that the order of the words that are added is preserved.